Transport behaviour

sasa1977 opened the pull request:
This is a first milestone in the discussed transport simplification. 

Implemented:
- Basic behaviour (`Phoenix.Transports.Driver`)
- Unified socket driver ('Phoenix.Socket.Driver')
- Some refactoring (mostly notably moving the code out `Phoenix.Socket.Transport`)
- Integration test for the socket driver (mostly a duplication of the existing websocket test, but independent from the transport)
- Some documentation

Still missing:
1. Making the driver implementation configurable (currently it's hardcoded)
2. Additional documentation & examples
3. Some final cleanup
4. Renaming/moving files to proper places

The point 1 is especially important, and I currently don't have a clear idea on how to do it. The problem is that currently transports are defined in the socket module. But the socket is now a construct of only one specific driver. Other drivers will not even know of sockets, so we need some other place to hold configurations. However, since we don't want to break the existing API (where transports are defined in sockets), it's not obvious to me how can we do it.

sasa1977 committed:
dialogue behaviour

sasa1977 committed:
channel driver

sasa1977 committed:
dialogue -> Phoenix.Transports.Driver

sasa1977 committed:
move the protocol version to socket driver

sasa1977 committed:
move transport behaviour to Phoenix.Transport

sasa1977 committed:
rename handler to socket_handler

handler is a too generic name, which makes the code harder to follow.

socket_handler also serves as an indication of couplings. Before we reach the socket driver, we shouldn't rely on sockets at all.

sasa1977 committed:
unify common plug operations

sasa1977 committed:
long polling through channel driver

sasa1977 committed:
move channel driver specifics out of Phoenix.Socket.Transport

sasa1977 committed:
tests for socket driver

sasa1977 committed:
remove :socket_push tag from the websocket serializer

sasa1977 committed:
websocket through channel driver

sasa1977 committed:
Phoenix.Channel.Driver -> Phoenix.Socket.Driver

sasa1977 committed:
change init params

Since the transport doesn't know what the implementation above needs, we need to include the union for all known cases. Current parameters are all that's needed in the only existing implementation.

sasa1977 committed:
consistent channel -> socket messages

Previously each serializer determined the message format which introduced a strong coupling between the serializer and the socket process. The shape of messages arriving to the socket process depended on the serializer (which is transport related).

Using consistent message format allows us to handle messages to the socket process in the same way, regardless of the encoder/fastlaner/transport which is used.

sasa1977 committed:
remove the terminate callback

sasa1977 committed:
documentation

sasa1977 commented:
@josevalim @chrismccord 

Here's the first pull which implements (hopefully) most of the features. As mentioned, this PR happens in this fork, it goes from one branch to another. I added you as collaborators to this repo. Once we have everything working, I can make the PR from the target branch to Phoenix master.

You can see some notes on what's done and what's missing in the PR description. There are a lot of changes, so my suggestion is that you review individual commits. They tell a story of how the change happened, and break down the amount of code which must be reviewed. The central change happens in the first four commits, and after that it's mostly refactoring, tests, and docs.

Happy reviewing :-)

josevalim commented:
I will try to review it tomorrow. Sorry for the delay.

sasa1977 commented:
No problem and no hurry :-)

josevalim commented on lib/phoenix/channel/driver.ex:
```diff
@@ -0,0 +1,227 @@
+  powered by transports, such as websocket or long polling, while this module
+  provides the stock implementation of the protocol, with following properties:
+
+  - Each channel runs in a separate process.
```
Use `*` and indent the bullet with two spaces and each paragraph with four. :)

josevalim commented on lib/phoenix/channel/server.ex:
```diff
@@ -147,9 +147,9 @@ defmodule Phoenix.Channel.Server do
       when is_binary(topic) and is_map(payload) do
 
-    send pid, serializer.encode!(
+    send pid, {:channel_push, serializer.encode!(
```
Do you remember why we had to wrap it here? I know we discussed it but I don't remember why. :)

sasa1977 replied:
The message is already wrapped in the current master, but the wrapping is delegated to the serializer. So [websocket serializer adds a tag](https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/transports/websocket_serializer.ex#L25), while [longpoller uses a struct](https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/transports/long_poll_serializer.ex#L25).

Consequently, [I had to deal with both kinds of messages in the generic socket driver](https://github.com/sasa1977/phoenix/blob/46dcff61caec806373e6dddd2e8fa279c0f6d640/lib/phoenix/channel/driver.ex#L62-L69). Moreover, such driver wouldn't work with other transports, since they might use some other tagging mechanism.

So [I made this change](https://github.com/sasa1977/phoenix/commit/eced974e6069d5bf65dec0e8fd93952c08123867) and made channel server responsible for the tag of the pushed message.

josevalim replied:
Do you know why we even tag in the first place? I think it is because the driver process itself may receive Broadcast messages, right? So we tag to differentiate.

sasa1977 replied:
Yes, these are the messages sent out by the channel process to its socket, instructing it to push message/reply to the client. So the socket driver needs to handle these messages in a particular way, by producing the outgoing message. See [here](https://github.com/sasa1977/phoenix/blob/transport-behaviour/lib/phoenix/channel/driver.ex#L96-L98).

josevalim replied:
My question is why we don't simply send it without the wrapping, since the wrapper is removed anyway. But I think it probably makes sense to be explicit.

sasa1977 replied:
Yes, I suppose we could use non-wrapped messages, but I also prefer explicitness. If you insist, I'll remove the tag.

josevalim replied:
Let's keep the tag. I am just rationalizing it. :)

josevalim commented on lib/phoenix/transports/long_poll.ex:
```diff
@@ -62,18 +62,13 @@ defmodule Phoenix.Transports.LongPoll do
     conn
-    |> code_reload(opts, endpoint)
-    |> fetch_query_params
+    |> Transports.Utils.init_plug_conn(endpoint, socket_handler, transport, &status_json(&1, %{}))
```
I understand why you did it but I would like to avoid coupling. `init_plug_conn` does nothing and everything at the same time and I think it makes more sense to call the individual functions one by one.

josevalim commented on lib/phoenix/transports/websocket.ex:
```diff
@@ -50,29 +50,17 @@ defmodule Phoenix.Transports.WebSocket do
-            {:ok, conn, {__MODULE__, {socket, opts}}}
+    conn
+    |> Phoenix.Transports.Utils.init_plug_conn(endpoint, socket_handler, transport)
+    |> case do
```
We can use `with` here now!

sasa1977 replied:
Not sure how do you see this done with `with`? We still need to branch on `:halted` and handle both cases in a specific way.

josevalim replied:
I was thinking you would do:

```
with %{halted: false} = conn <- Utils.force_ssl,
     %{halted: false} = conn <- Utils.check_origin,
     do: dispatch(conn, ....)
```

If you look closely, dispatch, force_ssl and check_origin all check for halted. But it is alright, we don't need to change this as part of this PR. :)

sasa1977 replied:
Ah, nicely spotted! I'll try to play and see how it looks with `with`, it's a good chance to try it under my fingers :-)

sasa1977 replied:
I made a trial switch to `with` in 71f2594800f7c77d7dc82ab0a9f04b07ae65db47 (previously, I deduplicated the pipeline in fe7a29fd77ee957ff5f90089643b523416791185). Let me know what you think.

josevalim commented on lib/phoenix/channel/driver.ex:
```diff
@@ -0,0 +1,227 @@
+  @doc """
+  Returns the Channel Transport protocol version.
+  """
+  # TODO(sj): do we really need to expose this? It seems to be used only in tests.
```
I think it is nice for tools or anything that want to show the protocol version from the elixir side.

josevalim commented:
@sasa1977 this is beautiful. I have dropped some comments but they are minor, really.

Regarding 1, keep in mind the socket needs to export `__transport__` today: https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/endpoint/cowboy_handler.ex#L68

So I would say that, when we build `__transport__` inside Phoenix.Socket, we should automatically add a `:driver` option (which in this case defaults to `Phoenix.Socket.Driver`.  I would also store the `:socket` inside the `config`, alongside the driver, and pass the whole `config` to the transport on start. So this line:

https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/endpoint/cowboy_handler.ex#L73

Would be something like:

```
 {module, {endpoint, config}}
```

And both driver and socket can be found inside `config`. We no longer need the transport name nor to keep the `__transport__` functions around. I would also add `__transports__` to be part of the `Phoenix.Transports.Driver` API, except `Phoenix.Socket` implements it in a way it proxies to another module.

Thoughts?

sasa1977 commented:
@josevalim Not sure I quite understand. As I see [here](https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/endpoint/cowboy_handler.ex#L67), the endpoint deals directly with sockets.

We can't have that in the new implementation, because socket is an implementation detail of a particular driver. But we also can't know about transports without asking the socket about it.

We could make `socket path, some_module` macro to generate something like `{path, Phoenix.Socket.Driver, some_module}` and then we could fetch drivers instead of sockets:

``` elixir
for {path, driver, driver_opts} <- endpoint.__drivers__
```

Then, `Phoenix.Socket.Driver` would expose `transports/1` function which becomes the part of the `Phoenix.Transports.Driver` behaviour. Now, we could call the driver to supply transports, so the complete comprehension could look like:

``` elixir
for {path, driver, driver_options} <- endpoint.drivers,
    {transport, {module, config}} <- driver.transports(driver_options),
    # Allow handlers to be configured at the transport level
    handler = config[:cowboy] || default_for(module),
    do: {Path.join(path, Atom.to_string(transport)),
         handler,
         {module, {endpoint, driver, driver_options, transport}}}
```

Notice that in this proposal I need to include `driver_options`. Moreover, I'm still passing the transport (which is in fact the transport name, such as :websocket or :longpoll), since this transport_name is used throughout the code, and it's also a documented member of `%Phoenix.Socket{}`.

It's also worth noting that [the final path is always suffixed by the transport name](https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/endpoint/cowboy_handler.ex#L71). Perhaps we could make this configurable by moving the path shaping to `Phoenix.Transports.Driver.transports` which would have to return `{transport, path, {module, config}}`.

sasa1977 committed:
format the bullet list

sasa1977 committed:
deduplicate conn initialization

sasa1977 committed:
remove to-do comment

As clarified by José: "I think it is nice for tools or anything that want to show the protocol version from the elixir side."

josevalim commented:
@sasa1977 your proposal also looks great. I think though we are passing too much information around, some of that has to be redundant. We could probably get by with something like this:

```
for {path, driver, driver_options} <- endpoint.drivers,
    {path, module, config} <- driver.transports(path, driver_options),
    # Allow handlers to be configured at the transport level
    handler = config[:cowboy] || default_for(module),
    do: {path,
         handler,
         {module, {endpoint, driver, config}}}
```

The `transport_name` and `driver_options` can all be encoded in the `config` once `driver.transports(path, driver_options)` is called. We should likely rename `module` to `transport` in this case, I think it will be clearer.

What do you think?

sasa1977 commented:
@josevalim Yes, this looks very nice. With this we have one place where configuration is specified, and can possibly remove `__transport__` from user's socket modules. I'll take that direction.

One remaining question is how can we specify our own driver in the endpoint? I think we need another macro, or an overload of the `socket` macro. I'm somewhat more in favour of the former, because in this version I regard socket as channel specific.

sasa1977 committed:
use with in the conn setup pipeline

sasa1977 commented:
@josevalim We have one minor problem. Transports currently need to know the socket module so they can cache some actions per socket. In particular, [these calls](https://github.com/sasa1977/phoenix/blob/transport-behaviour/lib/phoenix/transports/websocket.ex#L62-L63) will internally invoke [this](https://github.com/sasa1977/phoenix/blob/transport-behaviour/lib/phoenix/transports/utils.ex#L47) and [this](https://github.com/sasa1977/phoenix/blob/transport-behaviour/lib/phoenix/transports/utils.ex#L120)

We need to make transports agnostic of socket handlers. One way is to require another function in the behaviour, for example `type_id/1` or something of the sort, and then the implementation can return e.g. the socket module.

Another possible direction is to consider moving these private functions (`force_ssl_config` and `check_origin_config`) to the code of `Phoenix.Socket.Driver.transports/1`. As far as I can tell, these functions do indeed deal with preparing the configuration, so we could execute them once while the driver config is being setup, and the config is properly passed to transports and the driver.

I tend to prefer the second option. It means we don't need to cache anything nor introduce another, possibly confusing function in the behaviour. The only downside I see is that custom driver will need to possibly copy-paste the code of these two functions. Given that writing a custom driver is not a common scenario, I don't think that would be a huge problem.

sasa1977 committed:
socket-agnostic transports and the driver behaviour

josevalim commented:
> I tend to prefer the second option. It means we don't need to cache anything nor introduce another, possibly confusing function in the behaviour. 

Sounds good to me. We could have force_ssl_config and check_origin_config functions as well that we would call when setting up transports.

sasa1977 commented:
> We could have force_ssl_config and check_origin_config functions as well that we would call when setting up transports.

Not sure what you mean by this. Do you suggest we export these functions in the utils module, so future transports can reuse them? Or is it something else?

josevalim commented:
Yes. So in driver.transports you call force_ssl_config and then you just call force_ssl later on

sasa1977 commented:
@josevalim I made the first draft of discussed ideas in af12cae9161324ed7b7ba69bb4c5e980030d8ee9

I introduced [another struct](https://github.com/sasa1977/phoenix/blob/af12cae9161324ed7b7ba69bb4c5e980030d8ee9/lib/phoenix/socket/dialogue.ex#L3-L26) for the `config` data I think this improves type safety, and can also serve as the documentation point.

In this pull you'll also find a few commits I pushed yesterday which address your other points (such as `with` refactoring).

Feel free to review, I won't touch the code until I get your feedback.

The remaining issues are:
1. Decide how will we allow users to specify their own drivers in the endpoint
2. Write docs and tests for custom drivers
3. Handle remaining to-do points (mostly docs & rename/move some files to proper places)
4. Merge Phoenix master, resolve conflicts, test with the latest implementation

Once that's done, I can make a PR to Phoenix master, so this internal PR doesn't even need to be merged in this repo (which I'll delete anyway once the work is merged to upstream).

josevalim commented on lib/phoenix/channel/driver.ex:
```diff
@@ -0,0 +1,252 @@
+    for {transport_name, {transport, transport_opts}} <- socket_handler.__transports__ do
+      {
+        Path.join(path, Atom.to_string(transport_name)),
+        %Transports.Driver.Config{
```
I am not a fan of the struct because it defeats precisely the point of returning just a map, which is to not impose shape. I would be fine if the struct was defined in the driver itself, but imposing it from the transport is a :-1: in my opinion.

josevalim replied:
To expand on that, if the driver doesn't care about the transport name, driver_opts and so on, it should not be forced to return that.

josevalim replied:
It also feels backwards we are processing force_ssl and check_origin for all transports, as it may not be required to all of them.

sasa1977 replied:
Well, I actually started with kw list, then moved to a map, then moved to a struct. The reason is precisely because these fields are used in various places in the transport layer, so they are in fact mandatory. The final point where the struct is passed is the driver. Admittedly, the driver might not need all of these fields, but IMO that's fine. It can take what it needs, and ignore the rest, so I don't think this is a problem.

Using a map (or kw list) removes the type safety we get with struct [here](https://github.com/sasa1977/phoenix/blob/transport-behaviour/lib/phoenix/socket/dialogue.ex#L66) and [here](https://github.com/sasa1977/phoenix/blob/transport-behaviour/lib/phoenix/socket/dialogue.ex#L9-L15). It becomes less obvious what the concrete driver needs to return, and we basically depend on convention.

One option is to simply "debundle" mandatory args and pass them on individually, but personally I think this greatly reduces the reading experience. That's exactly what the code used to look like earlier, and personally I had a lot of trouble understanding what it does. Reading through various functions which pass around 5-6 params is very demotivating.

josevalim replied:
> That's exactly what the code used to look like earlier, and personally I had a lot of trouble understanding what it does. 

My point is that this problem was not fixed, it was just "made official" by putting everything in a struct. It still feels I am passing too much around and that I am being forced to do a bunch of bookkeeping that I don't actually care about. In particular:
1. `driver_opts` should not be needed downstream. We should process it all at this function
2. I don't understand why `transport_name` has a specific field imposed to all transports. Isn't it a Socket.Driver thing?

So I buy in on only two fields in the struct (transport and options) which IMO would then be better served by a tuple. 

You could argue that both driver_opts and transport_opts are needed but, if that is the case, I would like to see the driver receiving _only_ the `driver_opts`. If the driver is receiving everything, it is all coupled anyway, so I would just put everything in kw (for example, today the driver is accessing the transport serializer).

sasa1977 replied:
OK, so here's the breakdown.

Transport layer (including the cowboy handler) requires:
- full path
- transport module
- transport options

The socket driver `transports` function needs:
- endpoint
- socket module
- base path

The socket driver `init` function needs:
- socket module
- serializer module
- transport name
- transport module

I'd like to see the mandatory stuff somehow mandated, rather than letting them be defined by conventions and relying on `fetch!`. 

We could thus, separate driver options, driver init options, and transport options. In other words, the `transport` callback would return `[{path, transport_module, transport_opts, driver_init_opts}]`, and these would be then passed through the transport layer. The driver `init` callback would receive `endpoint`, `driver_init_opts`, and `params`.

That way we're more explicit about what's mandated while there's no coupling of unrelated data in the same struct. What do you think?

josevalim replied:
Both `socket_module` and `transport_name` _are_ specific to this driver implementation. Since your work is exactly in abstracting them, I don't think they should be required. Here is my proposal:

```
@spec driver.transports(endpoint, path, opts) ::
  {path, transport_module, transport_options}
```

The only reason to return `driver_opts` is if the transport will now explicitly pass a subset of its options to driver, if we are going to support that, then I am OK with returning `{path, transport_module, transport_opts, driver_opts}`. If we support `driver_opts`, we don't need to pass the endpoint to the driver because it can be added to `driver_opts`.

sasa1977 replied:
Well, yeah, in my proposal `socket_module` and `transport_name` would be included in `driver_init_opts`, so the driver chooses which data will it get later in the `init` callback.

Also, the driver already knows all the transport options, because it's the one fetching these options from the transport module and passing them to the rest of the system through the `transports` callback. So at this point, the driver can fetch all the transport options it will need later and stash them into `driver_init_opts`.

I guess in your proposal you'd bundle it all into `transport_opts` and pass them to the `init`? I guess that's fine too, but then it seems to me that the same kind of bundling happens which you argued against, no? Because now we have some transport options and driver options stashed together. This might even lead to some accidental bugs, if a driver and a transport select the same atom for some key.

josevalim replied:
Ok, to make life easier, let's go with a single set of options instead of splitting into transport and driver. We don't need to pass the endpoint when starting the transport nor the driver because it shuold be added to the options (alongside the transport name).

josevalim replied:
>  This might even lead to some accidental bugs, if a driver and a transport select the same atom for some key.

Yes but that is unlikely to happen. The benefit is that we have a single place to customize both of them and we don't  need to worry about splitting everything. Anyway, ignore my previous comment. We are aware of the trade-offs, your choice.

sasa1977 replied:
OK, actually it's true that most of what we're currently passing is related to transports, so I guess bundling it into transport_opts is fine for the moment. 

Just to be clear I understand your idea completely: `transport_name`, `socket_module`, and `endpoint` will end up in `transport_opts`, and we'll pass that all the way to the driver's `init`. And `transport_opts` remains a k-w list (since it's originally implemented as such). Is that right?

josevalim replied:
It can be a map. We just need to be sure `transport.config` also returns a map.

sasa1977 replied:
Are you sure about that? I mean, currently transports and users (through the `transport` macro) provide initial options as k-w lists. If we convert that to map later on, then we'd overwrite multiple values per key, right?

josevalim replied:
We are going to change transports now anyway, so we can always work with maps there. Then we just need to convert the user options into a map before calling the transport. No?

sasa1977 replied:
Well, yes we can of course always change that :-) I was simply operating under the assumption it will remain as it is. 

So, the default transport options become a map, user provided options must not contain duplicate keys, and the driver can add its own data into this map, and that will be passed around to transports and the driver's `init`. That's the idea?

josevalim replied:
Sounds great. :+1:

sasa1977 replied:
I only had time to address this issue (+ the associated `transport.config` contract change) in 8fe4add0851be5bd919a186c6f55f1af11abcf29. Let me know what you think

josevalim commented on lib/phoenix/socket.ex:
```diff
@@ -290,13 +280,13 @@ defmodule Phoenix.Socket do
 
   @doc false
-  def __transport__(transports, name, module, config) do
+  def merge_transport_spec(transports, name, module, config) do
```
If you remove the underscores, it will now be imported on `import Phoenix.Socket`. Please add them back or move the function elsewhere.

josevalim commented on test/phoenix/integration/socket_driver_test.exs:
```diff
@@ -0,0 +1,301 @@
+defmodule Phoenix.Integration.SocketDriverTest do
```
In Phoenix suite, we call integration tests the ones that go through an actual client like HTTP or Websocket. This is a regular test, so I would place it under `test/phoenix/socket/driver_test.exs`.

josevalim commented on test/phoenix/integration/socket_driver_test.exs:
```diff
@@ -0,0 +1,301 @@
+  test "joining an unmatched channel" do
+    {:ok, state} = init_driver()
+
+    assert  {
```
It feels this indention is uncommon on Phoenix source. My proposal is to break it in two parts:

``` elixir
assert {:error, :unmatched_topic, [%Reply{} = reply], _state} = join(...)
assert %{topic: ....} = reply
```

We should fix the other examples through the test source.

josevalim commented:
Thanks @sasa1977! I have added some comments. I am not a fan of the struct. I would rather return `{path, transport, config}` for each transport. `config` should be something like `[transport_name: ..., socket_handler: socket_handler] ++ transports_opts`. I.e. I don't want to impose any shape in the config.

Also, it is a little bit weird we are preparing check_origin and force_ssl in the socket driver while they are transport behaviour. I think the correct fix would be to [change the default config](https://github.com/sasa1977/phoenix/pull/1/files#diff-766b8936169a73d4d05015b285641364R27) in the transport to `config` and give it the transport config. Something instead of [this line](https://github.com/sasa1977/phoenix/blob/transport-behaviour/lib/phoenix/socket.ex#L290), we would call the following inside the new `transports` function:

``` elixir
transport.config(user_config)
```

And the transport would do the merging and the whole force_ssl and check_origin preparation.

Btw, Very nice work on the tests. Do you think we could remove some of the actual integration tests now? Or you think doing so would remove some of our confidence in the test suite?

sasa1977 commented:
Re tests, I'll go through them carefully and see if something can be removed, but I'd like to be careful here, even if it means keeping some duplicates. Ultimately, it's the integration tests which really verify end-to-end behaviour. In some cases, they also check some additional stuff. For example, many long polling tests verify the response status - something which can't be done in the socket driver test.

sasa1977 committed:
remove driver config struct

Also:

- use maps for config
- default_config/0 -> config/2

sasa1977 committed:
move socket tests to test/phoenix/socket

sasa1977 committed:
fix non-idiomatic indentation in tests

sasa1977 committed:
add underscores back to the function name

sasa1977 committed:
remove a few redundant tests

sasa1977 commented:
@josevalim I think I addressed everything from your second round of feedback.
Regarding tests, I only removed two tests from websockets. I'd say the remaining ones test the full communication path, so I'd rather leave them.

Remaining work from [this comment](https://github.com/sasa1977/phoenix/pull/1#issuecomment-189852556) still has to be done. In particular, we need to decide how will we allow people to specify custom drivers in the endpoint.

spscream commented:
@josevalim, @sasa1977, what about this work, any updates?

I'm currently implementing channels encrypted by libsodium on transport level and this is not obvious how to deal with custom transports if you need to override only specific cases from default transport.

josevalim commented:
Still on my list of things to do. But I think your work can be done with the serializers configuration provided by Phoenix. Also, there is no need to ping us in two places. :)

spscream commented:
@josevalim, yeah, sorry :) 

In my case serializers doesn't make sense, since I still using default serializer and I don't want to change application level payload, requirements are to encrypt all websocket frames, so I decide to put small wrapper between websocket handler and default Transports.Websocket implementation sitting on ws_init/ws_handle/etc.

josevalim commented:
I still think that replacing the whole transport seems to be unnecessary. In the same way you are still using the default serializer, you will quite likely still use the default transport. You should rather provide your own transport that uses another transport.

We should expose all frames via the serializer, if we don't, let us know and we can discuss extending the serializer API.

spscream commented:
@josevalim, providing my own transport which uses Phoenix.Transport.WebSocket was the easiest way to deal the problem, but may be I'm doing it wrong? 
I created custom transport, added defdelegate's to default implementation and put custom logic only in some functions. It was surprise what ws_init of my implementation wasn't called :)

I had to do several things:
- initialize key exchange with sending session pub key to user using event in custom topic;
- put initialized session encryption keys to the socket state;
- provide generated keys to serializer, so it could encrypt frames.

Last thing on this list with current implementation of serializers was possible only with inserting current session keys to every reply payload, which is lead to many changes in our codebase and I think application layer shouldn't worry about initializing and passing transport encryption keys :)
First thing on the list also wasn't easy to do - I didn't figure out how to send any info to user on socket connect.

It will be great If you extend api of serializers to access state of user socket, or allow to pass params in some way to it.

josevalim commented:
I see, that clarifies why a custom transport works best. Thanks for correcting me!

I will prioritize this work once Phoenix v1.3 is out.
